test/libsolidity/astPropertyTests/forLoop/isSimpleCounterLoop.sol

```diff
@@ -0,0 +1,29 @@
+contract C {
+  function f() public pure {
+     /// PositiveCase1: isSimpleCounterLoop
+     for(uint i = 0; i < 42; ++i) {
+     }
+     /// PositiveCase2: isSimpleCounterLoop
+     for(int i = 0; i < 42; i++) {
+     }
+     /// NegativeCase1: isSimpleCounterLoop
+     for(uint i = 0; i < 42; i = i + 1) {
+     }
+     /// NegativeCase2: isSimpleCounterLoop
+     for(uint i = 42; i > 0; --i) {
+     }
+     /// NegativeCase3: isSimpleCounterLoop
+     for(uint i = 42; i > 0; i--) {
+     }
+     /// NegativeCase4: isSimpleCounterLoop
+     for(uint i = 1; i < 42; i = i * 2) {
+     }
+  }
```

Comment:
There's further kinds of negative test cases we need here. At least:
- A proper ``for(uint i = 0; i < 42; ++i)``, but with several kinds of writes to the counter variable in the loop body (like ``i++`` or ``i = 43;``).
- Same as above, but with a write to the counter variable in inline assembly. Note that in the current state of the PR the result will be incorrect, so this still actually needs fixing.

Reply:
We should really test all the corner cases which may disable the simple counter, these are really simple to test now with AST property tests:
- user-defined `++` or `<` operators.
- empty loop expression
- complex condition or condition with `<=`
- condition LHS not being a variable name or not an integer variable
- different common type in the condition
- non-integer counter

Also, would be good to make sure that some things indeed do not bypass the optimization:
- variable as a condition RHS
- array length as a condition RHS

Reply:
I added all suggested tests, although with some remarks for:
 * user-defined `++` or `<` operators. -> Operator `++` cannot be user-defined.
 * non-integer counter -> I don't see in the docs a non-integer type of counter that would have both `<` and `++` operator... I used `address` which has `<`, but not `++`.
 * Counter being written in inline assembly -> Currently assignments are detected. But I am not sure if there are other ways to write to the counter. I guess one could play with `mstore` for example?